MaxGaranin/nodejs-aws-be
# Task 9 - BFF

### Task 9
Task description: https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task9-bff/task.md
Total: 6 / 7

## Links and examples
1. To add product use POST query with any of links:
Lambda API Gateway: https://vgu8neei3m.execute-api.eu-west-1.amazonaws.com/dev/products
Elastic Beanstalk BFF: http://maxgaranin-bff-api-dev.eu-west-1.elasticbeanstalk.com/products
No authorization headers needed.
Body:
{
    "title": "someTitle",
    "description": "someDescription",
    "price": 12,
    "count": 2
}

![image](https://user-images.githubusercontent.com/7194476/102054330-ac822880-3e02-11eb-8f9e-9ec3cffed7cd.png)

2. To work with CART use RS_CART_API with any of links:
Elastic Beanstalk API: http://maxgaranin-cart-api-dev.eu-west-1.elasticbeanstalk.com/
Elastic Beanstalk BFF: http://maxgaranin-bff-api-dev.eu-west-1.elasticbeanstalk.com/

- ping: http://maxgaranin-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart/ping

- login: http://maxgaranin-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart/api/login
Body:
{
    "username": "gitHubAcc",
    "password": "TEST_PASSWORD"
} 
You will get authorization token:
{
    "statusCode": 200,
    "message": "OK",
    "data": {
        "token_type": "Basic",
        "access_token": "Z2l0SHViQWNjOlRFU1RfUEFTU1dPUkQ="
    }
}

- update cart
http://maxgaranin-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart/api/profile/cart
Headers: Authorization - Basic Z2l0SHViQWNjOlRFU1RfUEFTU1dPUkQ=
Body:
{
    "items": [
        {
            "product": {
                "id": "id1",
                "title": "id1Title",
                "description": "description",
                "price": 1234.5
            },
            "count": 3
        }
    ]
}

![image](https://user-images.githubusercontent.com/7194476/102056129-5367c400-3e05-11eb-888d-eadc5b11a6c0.png)

## Evaluation criteria (each mark includes previous mark criteria)
* ( + ) **3** - A working and correct **express** application should be in the **bff-service** folder. Reviewer can start this application locally with any valid configuration in the **.env** file and this application should works as described in the task 9.1
* ( + ) **5** - The **bff-service** should be deployed with Elastic Beanstalk. The **bff-service** call should be redirected to the appropriate service : **product-service** or **CART**. The response from the **bff-service** should be the same as if **product-service** or **CART** services were called directly.

## Additional (optional) tasks
* ( + ) **+1** - Add a cache at the **bff-service** level for a request to the **getProductsList** function of the **product-service**. The cache should expire in 2 minutes.  
How to test:
  * Get products list
  * Create new product
  * Get products list - result shouldn’t have new product
  * Wait more than 2 minutes
  * Get products list - result should have new product
* ( - ) **+1** - Use **NestJS** to create **bff-service** instead of **express**
